### DISCUSSION TITLE
Global constants vs associated constants

### DISCUSSION THREAD

**Discussion by EgorKulikov** [near/nearcore#3992] [General] [answered]
I noticed that through our codebase we mostly use global constants. Since Rust 2017 there are [associated constants](https://doc.rust-lang.org/edition-guide/rust-2018/trait-system/associated-constants.html), which work kinda like public static final members in Java classes. I think they offer better structuring and less cluttering of global namespace. I propose to use them moving forward when constant you want to introduce have some clear type logically associated with it, and, probably, even refactor older constants that are similar in the same go

What do you think?

**Comment by bowenwang1996** [comment]
@frol @matklad

**Comment by frol** [comment]
I am all for the better code, and associated constants are a great way to structure the related entities. Let's use them where applicable :+1:

**Comment by matklad** [comment, marked as answer]
:+1:, with a caveat.

Caveat: Rust's primary units of encapsulation are a module and a crate, so generally it's perfectly fine to have stuff in module namespace. It is not a problem by itself. Often, the opposite is true -- people needlessly put stuff in impl blocks. 

So, if there's a bunch of `FOO_XXX` constants and a `Foo` type, it makes sense to write `Foo::XXX` instead. If, of the other hand, there's no single type to which a constant *clearly* belongs, it's better to leave it at the module level.
